File: android/app/src/main/java/com/podswitch/core/SwitchEngine.kt
package com.podswitch.core

/** Pure, side-effect-free decision engine mapping a [SwitchEvent] to a [SwitchAction]. */
object SwitchEngine {

    fun decide(
        event: SwitchEvent,
        config: Config,
        status: DeviceStatus,
    ): SwitchAction {
        if (!config.enabled) return SwitchAction.None
        if (config.targetDeviceId == null) return SwitchAction.None

        return when (event) {
            is SwitchEvent.AudioStarted -> {
                if (event.category !in config.enabledCategories) return SwitchAction.None
                if (!status.targetPaired) return SwitchAction.None
                if (status.targetActiveOutput) return SwitchAction.None
                // Coordination layer (authoritative): a peer device is actively playing on the
                // target right now — protect it, don't steal.
                if (config.yieldToOtherSource && status.peerActiveOnTarget) return SwitchAction.None
                // Reactive fallback: we lost the target to another source and haven't got it back.
                if (config.yieldToOtherSource && status.targetYielded) return SwitchAction.None
                when (config.mode) {
                    Mode.STEAL -> SwitchAction.Connect
                    Mode.ASK -> if (status.notificationPending) SwitchAction.None else SwitchAction.Notify
                }
            }

            // Explicit user request overrides the yield guard.
            SwitchEvent.UserAcceptedSwitch -> {
                if (!status.targetPaired) return SwitchAction.None
                if (status.targetActiveOutput) return SwitchAction.None
                SwitchAction.Connect
            }

            SwitchEvent.AudioStopped -> SwitchAction.None

            // Pure state input for the Coordinator; never an action on its own.
            is SwitchEvent.TargetConnectionChanged -> SwitchAction.None
        }
    }
}

File: windows/PodSwitch.Core/SwitchEngine.cs
namespace PodSwitch.Core;

/// <summary>Pure, side-effect-free decision engine mapping a switch event to an action.</summary>
public static class SwitchEngine
{
    public static SwitchAction Decide(SwitchEvent ev, Config config, DeviceStatus status)
    {
        if (!config.Enabled) return SwitchAction.None;
        if (config.TargetDeviceId is null) return SwitchAction.None;

        switch (ev)
        {
            case SwitchEvent.AudioStarted started:
                if (!config.EnabledCategories.Contains(started.Category)) return SwitchAction.None;
                if (!status.TargetPaired) return SwitchAction.None;
                if (status.TargetActiveOutput) return SwitchAction.None;
                return config.Mode switch
                {
                    Mode.Steal => SwitchAction.Connect,
                    Mode.Ask => status.NotificationPending ? SwitchAction.None : SwitchAction.Notify,
                    _ => SwitchAction.None,
                };

            case SwitchEvent.UserAcceptedSwitch:
                if (!status.TargetPaired) return SwitchAction.None;
                if (status.TargetActiveOutput) return SwitchAction.None;
                return SwitchAction.Connect;

            case SwitchEvent.AudioStopped:
                return SwitchAction.None;

            default:
                return SwitchAction.None;
        }
    }
}
